src/hotspot/share/gc/g1/g1BarrierSet.cpp

```diff
@@ -107,22 +107,24 @@ void G1BarrierSet::invalidate(MemRegion mr) {
   }
   volatile CardValue* byte = _card_table->byte_for(mr.start());
   CardValue* last_byte = _card_table->byte_for(mr.last());
-  // skip initial young cards
-  for (; byte <= last_byte && *byte == G1CardTable::g1_young_card_val(); byte++);
 
-  if (byte <= last_byte) {
-    OrderAccess::storeload();
-    // Enqueue if necessary.
-    Thread* thr = Thread::current();
-    G1DirtyCardQueueSet& qset = G1BarrierSet::dirty_card_queue_set();
-    G1DirtyCardQueue& queue = G1ThreadLocalData::dirty_card_queue(thr);
-    for (; byte <= last_byte; byte++) {
-      CardValue bv = *byte;
-      if ((bv != G1CardTable::g1_young_card_val()) &&
-          (bv != G1CardTable::dirty_card_val())) {
-        *byte = G1CardTable::dirty_card_val();
-        qset.enqueue(queue, byte);
-      }
+  // skip young gen cards
+  if (*byte == G1CardTable::g1_young_card_val()) {
+    assert(mr.word_size() <= HeapRegion::GrainWords, "Invalidated MemRegion is larger than HeapRegion in young gen");
```

Comment:
I would change this to a check that the memregion (in young) does not span multiple regions; for the same reason one could check that it is not humongous(is that true?), but I'm not sure what the background for this check is, so.
Feel free to improve the check.
Other than that comment, looks good.